group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -856,7 +899,8 @@ public boolean cleanupExpiredOffsets(String groupId, List<Record> records) {
         });
         metrics.record(OFFSET_EXPIRED_SENSOR_NAME, expiredPartitions.size());
 
-        return allOffsetsExpired.get();
+        // We don't want to remove the group if there are ongoing transactions.
+        return allOffsetsExpired.get() && !openTransactionsByGroup.containsKey(groupId);
```

Comment:
Just want to understand here, if there are ongoing transactions, `hasPendingTransactionalOffsets(groupId, topic, partition)` will be true and `allOffsetsExpired` will be set to false in L888. Why `!openTransactionsByGroup.containsKey(groupId)` is needed?

Reply:
Because there may be pending transactional offsets for partitions which are not in the main storage yet. When we iterate over `offsetsByTopic`, we don't see those. Therefore, it is better to prevent expiring the group if there are ongoing transactions.